subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultMultiRequestWorkerProcessBuilder.java

```diff
@@ -150,11 +150,14 @@ public MultiRequestClient<IN, OUT> build() {
 
             @Override
             public WorkerProcess start() {
-                try {
-                    workerProcess.start();
-                } catch (Exception e) {
-                    throw WorkerProcessException.runFailed(getBaseName(), e);
-                }
+                // Temporarily un-set the current build operation, since it shouldn't be leaked to the worker management code.
+                CurrentBuildOperationRef.instance().with(null, () -> {
```

Comment:
It's not 100% clear to me why this is necessary.  Doesn't this mean any log messages we generate while starting a worker process (e.g. via stdout/stderr) will not be associated with the task?

Reply:
This is necessary to prevent leaking the expensive JavaCompile operation, and yes, the logs for startup will likely no longer be associated with a task, but I don't feel like they should be? Those feel related to the Gradle daemon as a whole. I suppose it could be used to see part of why a task is taking so long, but there's already an issue with doing it this way -- when I tested with Checkstyle it didn't have an operation already, for reasons I'm not sure about. So these logs haven't been associated with most worker-based tasks for a while already.

If we want startup to be associated, I feel it should be done in a more managed way, where after startup it is dropped from the worker thread.

Reply:
Yeah, I can see the argument that the startup of the worker should not necessarily be a part of the task.  I think what I'm worried about is a worker failing to start because, I don't know, a bad argument is passed to the java executable, so the task fails, but in the build scan, when you say show in console log, it doesn't show the error because the message isn't associated with the task.  If you look at the full console log, you'll see the error message, but in a large log it might be difficult to find it.

I wonder if it's worth trying to drop the reference to the task operation from the stream handlers once the startup is complete and the response protocol is connected?  IOW, once the worker is started, any further output on those stream handlers is just treated like any other output.  Everything should be arriving on the response protocol at that point anyways.

I feel like this is something we could solve in a followup change if we wanted to - we don't necessarily need to address it in this change - but I think it's worth thinking about how we can make a startup error easy to diagnose in a build scan.  There might be other ways to approach it, too, like capturing worker logs as their own thing and just having the task know what workers it used and a way to navigate to the worker logs from the task in a build scan.

Reply:
Can we add a test that demonstrates this? This is enough to do it when there are Java sources:
```
tasks.withType(JavaCompile) {
   options.fork = true
   options.forkOptions.memoryInitialSize="-1"
}
```

Here's what we did in 8.0.2:
https://scans.gradle.com/s/g4teaapteivda/console-log?task=:compileJava

Here's what we do now (with this PR):
https://scans.gradle.com/s/dawgj4weasw2a/console-log?task=:compileJava

I don't think we should regress with this in 8.3.

Reply:
Looking around, I think there's something deeper going on. 

There are three paths for logging to end up associated with the task:
1. Errors when first starting the process. The current change breaks this association. 
2. Logging as it comes from the worker. I think the worker protocol takes care of this.
3. Unexpected failures _after_ startup, but not through the worker protocol.

(1) we can fix in a follow-up PR. 
(2) seems to be the easiest to understand and seems to be OK. The current operation is connected to the request, so it naturally changes when a worker is reused.
(3) I'm not sure what happens today or with this PR. On master, it seems like if the compiler daemon produced an error after the task finished, we would produce a build operation "at the wrong time" because it would be after the task action, but it would still be a child of the task action. With this PR, I think it would be unassociated with any build operation like (1).

There are a couple of things that look smelly with the implementation of the build operation -> logging to me. We have to explicitly wrap runnables in several related places. Incidentally, clearing the current build operation when starting the worker also removes the current build operation for the stream forwarding threads that live forever.

I have a couple of rough ideas/questions:
1. Can the ExecHandle/Runner forward start-up errors to the worker builder so that the exec infrastructure doesn't need to know about the current build operation?
2. Instead of using `CurrentBuildOperationPreservingRunnable` in multiple places, could the executor itself wrap things when it knows it needs to? 
3. Can the stream forwarders be dynamically given the current build operation?
4. Should we disconnect the stream forwarders between requests?

Reply:
> Unexpected failures after startup, but not through the worker protocol.

Under what circumstances do you think this might happen?  I believe once we call `LoggingManager.captureSystemSources()` we [disconnect the standard output and standard error streams](https://github.com/gradle/gradle/blob/ca9e8df9baa28648f7c1ba5cdf9148a434204300/subprojects/worker-processes/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java#L175) and only send back output/error events on the worker logging protocol we set up [here](https://github.com/gradle/gradle/blob/ca9e8df9baa28648f7c1ba5cdf9148a434204300/subprojects/worker-processes/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java#L157).  I don't think anything will arrive from the stream handlers at that point - it will all occur over the worker protocol.

I suspect what you mean is outside of a work request/response, in which case I think the log event would arrive without a build operation associated with it.  I think it may have always done this.

> Instead of using CurrentBuildOperationPreservingRunnable in multiple places, could the executor itself wrap things when it knows it needs to?

This seems like a good idea.  I wonder if `DefaultExecActionFactory.BuilderImpl` could just take an argument that just wraps the executor in something that decorates runnables with `CurrentBuildOperationPreservingRunnable`?

> Can the stream forwarders be dynamically given the current build operation?

I wonder if this is necessary.  Like I said, I don't think these get used again once the worker has started up and connected all of the logging stuff.  If so, I think we just need to determine what build operation to associate with it during startup (if any).

> Should we disconnect the stream forwarders between requests?

If I'm right about not using them anymore, we could probably just disconnect them after startup.

Reply:
> I believe once we call `LoggingManager.captureSystemSources()` we [disconnect the standard output and standard error streams](https://github.com/gradle/gradle/blob/ca9e8df9baa28648f7c1ba5cdf9148a434204300/subprojects/worker-processes/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java#L175)

OK, that seems to cover the case where the communication protocol is still working.

I was thinking of scenarios like:
- Work action is "evil" and starts its own thread that produces output. What do we do after the request is finished?
- Worker process is killed by OOM-killer and the JVM produces stderr. Where does that go?
- Gradle code unintentionally emits logging outside of a request. Where does that go?

> I suspect what you mean is outside of a work request/response, in which case I think the log event would arrive without a build operation associated with it.

I was thinking about this happening outside of a work request.  Does the log event get associated with any build operation (the root most one?) once on the daemon side?

>  I wonder if DefaultExecActionFactory.BuilderImpl could just take an argument that just wraps the executor in something that decorates runnables with CurrentBuildOperationPreservingRunnable?

Yeah, I was thinking something like that.  The executor would know that all threads need to be wrapped and requesting a "non-wrapped" executor would be special.

> Like I said, I don't think these get used again once the worker has started up and connected all of the logging stuff.

There's this existing daemon flag: 
https://github.com/gradle/gradle/blob/a138710ab109e271b17ee48dc49b359284cd3f66/subprojects/core/src/main/java/org/gradle/process/internal/ExecHandleRunner.java#L78

This looks suspiciously close to what we want:
1. Associate start-up messages to the build operation (so revert the changes this thread is on)
2. Once started, stdout/stderr is routed through something else
3. Stop stream threads 

This might also eliminate the extra threads we create.

Reply:
To answer some of the questions posed above:

> Work action is "evil" and starts its own thread that produces output. What do we do after the request is finished?
It goes to the output stream which is backed by the `WorkerLogEventListener` which pumps it back to the build process without a build operation.

> Gradle code unintentionally emits logging outside of a request. Where does that go?
Same.

> Worker process is killed by OOM-killer and the JVM produces stderr. Where does that go?
Same, although I think there is a race condition there where the connection thread may get stopped before the message is transmitted.

> I was thinking about this happening outside of a work request. Does the log event get associated with any build operation (the root most one?) once on the daemon side?
Pretty sure it arrives from the worker process without a build operation.  At that point, I think the build process then attaches it to the root build operation.